orttraining/orttraining/training_ops/cuda/activation/bias_gelu_grad_impl.cu

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cuda/activation/bias_gelu_grad_impl.h"
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "core/providers/cuda/shared_inc/fast_divmod.h"
+#include "orttraining/training_ops/cpu/activation/gelu_computation_mode.h"
+#include "orttraining/training_ops/cuda/activation/gelu_grad_impl_common.cuh"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T, typename GeluComputationMode, int num_consecutive_elements_per_group, int num_groups_per_thread>
+__global__ void BiasGeluGradDxKernel(int64_t bias_size, const T* dY, const T* X, const T* B, T* dX) {
+  const int64_t input_base_idx = bias_size * blockIdx.x + num_consecutive_elements_per_group * threadIdx.x;
+  const int64_t bias_base_idx = num_consecutive_elements_per_group * threadIdx.x;
+  const int64_t group_stride = num_consecutive_elements_per_group * blockDim.x;
+
+#pragma unroll
+  for (int group_idx = 0; group_idx < num_groups_per_thread; ++group_idx) {
+#pragma unroll
+    for (int element_idx = 0; element_idx < num_consecutive_elements_per_group; ++element_idx) {
+      const auto offset = group_stride * group_idx + element_idx;
+      if (bias_base_idx + offset < bias_size) {
+        dX[input_base_idx + offset] = ComputeGeluGradScalar(
+            dY[input_base_idx + offset], X[input_base_idx + offset] + B[bias_base_idx + offset],
+            GeluComputationMode{});
+      }
+    }
+  }
+}
+
+template <typename T, typename GeluComputationMode>
+void LaunchBiasGeluGradDxKernel(
+    int64_t input_size, int64_t bias_size,
+    const T* dY, const T* X, const T* B, T* dX) {
+  // each block handles bias_size elements
+  // there are input_size / bias_size blocks
+  constexpr int num_consecutive_elements_per_group = 4;
+  constexpr int num_groups_per_thread = 4;
+
+  const auto num_threads_per_block = CeilDiv(bias_size, num_consecutive_elements_per_group * num_groups_per_thread);
```

Comment:
>num_threads_per_block [](start = 13, length = 21)

Please add a few test case for different bias_size
one with common bert's size 
one is not dividable by 4

Reply:
added tests to orttraining/orttraining/test/training_ops/cpu/activation/activation_op_test.cc

---
In reply to: [443775458](https://github.com/microsoft/onnxruntime/pull/4146#discussion_r443775458) [](ancestors = 443775458)